homeassistant/components/light/group.py

```diff
@@ -0,0 +1,280 @@
+
+    async def async_turn_on(self, **kwargs):
+        """Forward the turn_on command to all lights in the group."""
+        await self._async_send_message(SERVICE_TURN_ON, **kwargs)
+
+    async def async_turn_off(self, **kwargs):
+        """Forward the turn_off command to all lights in the group."""
+        await self._async_send_message(SERVICE_TURN_OFF, **kwargs)
+
+    async def async_update(self):
+        """Query all members and determine the group state."""
+        states = self._child_states()
+        on_states = [state for state in states if state.state == STATE_ON]
+
+        self._state = _determine_on_off_state(states)
+
+        self._brightness = _reduce_attribute(on_states, ATTR_BRIGHTNESS)
+
+        self._xy_color = _reduce_attribute(
+            on_states, ATTR_XY_COLOR, reduce=_mean_tuple)
+
+        self._rgb_color = _reduce_attribute(
+            on_states, ATTR_RGB_COLOR, reduce=_mean_tuple)
+        if self._rgb_color is not None:
+            self._rgb_color = tuple(map(int, self._rgb_color))
+
+        self._white_value = _reduce_attribute(on_states, ATTR_WHITE_VALUE)
+
+        self._color_temp = _reduce_attribute(on_states, ATTR_COLOR_TEMP)
+        self._min_mireds = _reduce_attribute(
+            states, ATTR_MIN_MIREDS, default=154, reduce=min)
+        self._max_mireds = _reduce_attribute(
+            states, ATTR_MAX_MIREDS, default=500, reduce=max)
+
+        self._effect_list = None
+        all_effect_lists = list(
+            _find_state_attributes(states, ATTR_EFFECT_LIST))
+        if all_effect_lists:
+            # Merge all effects from all effect_lists with a union merge.
+            self._effect_list = list(set().union(*all_effect_lists))
+
+        self._effect = None
+        all_effects = list(_find_state_attributes(on_states, ATTR_EFFECT))
+        if all_effects:
+            # Report the most common effect.
+            effects_count = Counter(itertools.chain(all_effects))
+            self._effect = effects_count.most_common(1)[0][0]
+
+        self._supported_features = 0
+        for support in _find_state_attributes(states, ATTR_SUPPORTED_FEATURES):
+            # Merge supported features by emulating support for every feature
+            # we find.
+            self._supported_features |= support
+        # Bitwise-and the supported features with the GroupedLight's features
+        # so that we don't break in the future when a new feature is added.
+        self._supported_features &= SUPPORT_GROUP_LIGHT
+
+    def _child_states(self) -> List[State]:
+        """The states that the group is tracking."""
+        states = [self.hass.states.get(x) for x in self._entity_ids]
+        return list(filter(None, states))
+
+
+def _find_state_attributes(states: List[State],
+                           key: str) -> Iterator[Any]:
+    """Find attributes with matching key from states."""
+    for state in states:
+        value = state.attributes.get(key)
+        if value is not None:
```

Comment:
If attributes is equal to `{ 'bla': None }`, it would not yield the value. Is this on purpose and would it ever make sense to return `None`?

Reply:
I think I once saw some platforms using state attributes as `None` to indicate that they didn't exist. I replaced the `key in state.attributes` check in the last commit so that we don't break the `_reduce_attribute` code if that happens.

Reply:
👍